### DISCUSSION TITLE
Interface Notifications

### DISCUSSION THREAD

**Discussion by anfss** [xlab-si/xopera-opera#214] [Q&A] [answered]
Hi,

Interface notifications are not working for me.
Can anyone share an example with notifications?

Thank you

**Comment by anzoman** [comment, marked as answer]
Hi, thanks for starting this discussion. 

Currently we don't have any solid example with interface notifications. This is mostly because [TOSCA](https://docs.oasis-open.org/tosca/TOSCA-Simple-Profile-YAML/v1.3/os/TOSCA-Simple-Profile-YAML-v1.3-os.html#_Toc26969437) is not describing this part in detail and gives just one small example, so we did not find any real use case for that. This means that xOpera doesn't do anything but just parse the interface notifications. We tried to use the interface notifications to allow scaling but we rather proceeded with with TOSCA triggers and policies.

![image](https://user-images.githubusercontent.com/41727816/123817425-70bf3400-d8f8-11eb-83ac-4fc30ada8700.png)

If you have any good example on how these notifications could be used you're welcome to explain it here and we can see if this is something that can be implemented within xOpera.

**Reply by anfss** [reply]
Thank you for your answer.
From what I understood from the standard, notifications seem to be a repetition of triggers but they have an additional functionality, they allow to receive information from external systems. For example, in the “Scale Up” and “Scale Down” examples it would allow to receive information for a more efficient “Scale” operation. I honestly don't quite understand the intent of the notification “implementation” field nor do I know how orchestrators can use this information.
By the way, if you allow me, I would like to know if there is any way, from a trigger, to get the instantiation of new nodes. We can have nodes that did not satisfy certain requirements and the operations of a trigger can allow this new instance.

**Reply by anzoman** [reply]
Many thanks for your view on this feature! @cankarm, @sstanovnik, @dradx, @alexmaslenn do you have something to add here or did you ever try to understand or use these TOSCA notifications?

**Reply by cankarm** [reply]
Thanks for initiating the discussion trying to clarify TOSCA standard ideas and features.

I never looked closely at the notifications and I would have the same questions as you, @anfss, regarding how to make the implementation part. For me, it is not clear if this would define a concrete monitoring agent(notification agent), that sends the notification to the orchestrator, or, maybe only the "receiving point and handling" for the notification at the orchestrator. 

Regarding the trigger.... I guess it has the implementation part? There you can define, how to set up a new node?

**Reply by anfss** [reply]
So, this means that the orchestrator doesn't try to instantiate new nodes (node templates) that become possible when the requirements become valid with new data that is introduced by external services (for example through notifications outputs).
The only way to have the installation of new services will be to imperatively execute scripts that make their configuration and execution, without this being reflected in the set of node templates managed by the orchestrator.

**Reply by anzoman** [reply]
Yes @anfss, the instantiation of new nodes (e.g. new VMs) at runtime along with TOSCA triggers is currently possible mostly through trigger operations (i.e., Ansible playbooks). We did not focus so much on deploying new nodes through TOSCA during runtime because the standard does not provide much examples on that.

So, basically when you run [`opera notify`](https://xlab-si.github.io/xopera-docs/opera_cli.html#opera-notify) CLI command the operations from triggers are invoked and there you can for instance perform various scaling actions. There is also another way - since TOSCA service templates specify a set of nodes that need to be instantiated at service deployment time, we have [`opera update`](https://xlab-si.github.io/xopera-docs/opera_cli.html#opera-update) CLI command, which lets you redeploy your service template based on the discovered changes in the template and in the instances. So, this brings a possibility to instantiate new nodes on TOSCA level. Please also check our [docs](https://xlab-si.github.io/xopera-docs/) and [examples](https://github.com/xlab-si/xopera-examples) and feel free to reach out again if you have any additional questions or suggestions.

**Reply by anfss** [reply]
Thank you. The "opera update" CLI command can be the solution for my problem.
Please consider to include notifications in a next version, it would be appreciated.
Once again, thank you very much.
